lib/c-api/src/wasm_c_api/mod.rs

```diff
@@ -1150,31 +1156,48 @@ pub struct wasm_ref_t;
 
 // opaque type which is a `RuntimeError`
 #[repr(C)]
-pub struct wasm_trap_t {}
+pub struct wasm_trap_t {
+    inner: RuntimeError,
+}
 
-#[no_mangle]
-pub unsafe extern "C" fn wasm_trap_delete(trap: Option<NonNull<wasm_trap_t>>) {
-    if let Some(t_inner) = trap {
-        let _ = Box::from_raw(t_inner.cast::<RuntimeError>().as_ptr());
+impl From<RuntimeError> for wasm_trap_t {
+    fn from(other: RuntimeError) -> Self {
+        Self { inner: other }
     }
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn wasm_trap_message(
-    trap: *const wasm_trap_t,
-    out_ptr: *mut wasm_byte_vec_t,
-) {
-    let re = &*(trap as *const RuntimeError);
-    // this code assumes no nul bytes appear in the message
-    let mut message = format!("{}\0", re);
-    message.shrink_to_fit();
+pub unsafe extern "C" fn wasm_trap_new(
+    _store: &mut wasm_store_t,
+    message: &wasm_message_t,
+) -> Option<Box<wasm_trap_t>> {
+    let message_bytes: &[u8] = message.into_slice()?;
+    let message_str = c_try!(std::str::from_utf8(message_bytes));
+    let runtime_error = RuntimeError::new(message_str);
+    let trap = runtime_error.into();
+
+    Some(Box::new(trap))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_trap_delete(_trap: Option<Box<wasm_trap_t>>) {}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_trap_message(trap: &wasm_trap_t, out_ptr: &mut wasm_byte_vec_t) {
+    let message = trap.inner.message();
+    out_ptr.size = message.len();
+    // TODO: make helper function for converting `Vec<T>` into owned `wasm_T_vec_t`
+    let mut dupe_data: Box<[u8]> = message.into_bytes().into_boxed_slice();
+    out_ptr.data = dupe_data.as_mut_ptr();
+    mem::forget(dupe_data);
```

Comment:
`Box::leak` has lifetime stuff associated it with it and will return a reference to a fat pointer, we want a word-sized pointer so we get a mut pointer from the slice itself and then prevent the box from being dropped.  If you mean calling `Box::leak` or `Box::into_raw` rather than `forget`, that's fine too, I like `forget` though because it's explicit that we're just preventing it from being Dropped and nothing else